src/script/interpreter.cpp

```diff
@@ -1322,6 +1378,75 @@ template void PrecomputedTransactionData::Init(const CMutableTransaction& txTo,
 template PrecomputedTransactionData::PrecomputedTransactionData(const CTransaction& txTo);
 template PrecomputedTransactionData::PrecomputedTransactionData(const CMutableTransaction& txTo);
 
+static const CHashWriter HASHER_TAPSIGHASH = TaggedHash("TapSighash");
+
+template<typename T>
+bool SignatureHashSchnorr(uint256& hash_out, const T& tx_to, uint32_t in_pos, uint8_t hash_type, SigVersion sigversion, const PrecomputedTransactionData& cache)
+{
+    uint8_t ext_flag;
+    switch (sigversion) {
+    case SigVersion::TAPROOT:
+        ext_flag = 0;
+        break;
+    default:
+        assert(false);
+    }
+    assert(in_pos < tx_to.vin.size());
+    assert(cache.m_bip341_taproot_ready && cache.m_spent_outputs_ready);
+
+    CHashWriter ss = HASHER_TAPSIGHASH;
+
+    // Epoch
+    static constexpr uint8_t EPOCH = 0;
+    ss << EPOCH;
+
+    // Hash type
+    const uint8_t output_type = (hash_type == SIGHASH_DEFAULT) ? SIGHASH_ALL : (hash_type & SIGHASH_OUTPUT_MASK); // Default (no sighash byte) is equivalent to SIGHASH_ALL
+    const uint8_t input_type = hash_type & SIGHASH_INPUT_MASK;
+    if (!(hash_type <= 0x03 || (hash_type >= 0x81 && hash_type <= 0x83))) return false;
+    ss << hash_type;
+
+    // Transaction level data
+    ss << tx_to.nVersion;
+    ss << tx_to.nLockTime;
+    if (input_type != SIGHASH_ANYONECANPAY) {
+        ss << cache.m_prevouts_single_hash;
+        ss << cache.m_spent_amounts_single_hash;
+        ss << cache.m_spent_scripts_single_hash;
+        ss << cache.m_sequences_single_hash;
+    }
+    if (output_type == SIGHASH_ALL) {
+        ss << cache.m_outputs_single_hash;
+    }
+
+    // Data about the input/prevout being spent
+    const auto& witstack = tx_to.vin[in_pos].scriptWitness.stack;
+    bool have_annex = witstack.size() > 1 && witstack.back().size() > 0 && witstack.back()[0] == ANNEX_TAG;
+    const uint8_t spend_type = (ext_flag << 1) + (have_annex ? 1 : 0); // The low bit indicates whether an annex is present.
+    ss << spend_type;
+    if (input_type == SIGHASH_ANYONECANPAY) {
+        ss << tx_to.vin[in_pos].prevout;
+        ss << cache.m_spent_outputs[in_pos];
+        ss << tx_to.vin[in_pos].nSequence;
+    } else {
+        ss << in_pos;
+    }
+    if (have_annex) {
+        ss << (CHashWriter(SER_GETHASH, 0) << witstack.back()).GetSHA256();
```

Comment:
So far as I can tell, the [ordering of serialization](https://github.com/jamesob/bitcoin/blob/301959fa7482cf749504864789e6f9f4470b8bb6/src/serialize.h#L238) is reversed here relative to what [the BIP](https://github.com/bitcoin/bips/blob/master/bip-0341.mediawiki#common-signature-message) says ("the SHA256 of *(compact_size(size of annex) || annex)*"). Existing serialization looks like it has the size coming after the annex itself. Even if I'm right, not sure how much this matters.

Reply:
I don't understand. Why would the length go after the data? Nothing serializes that way, and the functional tests would fail if that was the case.

Reply:
Based on the serialization code I link to, it looks as though the size is being written after the data itself (which is the reverse of what the BIP says), but maybe I'm missing something?

Reply:
That serialization code you link to:
* Is serializing a `Span<unsigned char>` as a fixed length object, without any length at all. It's invoking the `Stream` function `write(const unsigned char* data, size_t length)`, which writes the `length` bytes starting at `data`. It doesn't write the length itself.
* Isn't used here (witstack.back() is a `std::vector<unsigned char>`, not a `Span), which has different serialization code a bit further down the file.

Reply:
Hm yeah I guess I must be wrong because the tests encode in [the right order](https://github.com/bitcoin/bitcoin/pull/19953/commits/f06e6d03452cf5e0b1a0863afb08c9e6d3ef452e#diff-3f25b0e0fbc5dd36a62417827b38517c9c5bdbfc3f4787d5925527c344bda104R775) (per `ser_string`'s [definition](https://github.com/jamesob/bitcoin/blob/301959fa7482cf749504864789e6f9f4470b8bb6/test/functional/test_framework/messages.py#L106-L107)). Maybe I'm looking at the wrong serializer.